platform/rpc/nats/bus.go

```diff
@@ -0,0 +1,246 @@
+package nats
+
+import (
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/nats-io/nats"
+
+	"github.com/weaveworks/flux"
+	"github.com/weaveworks/flux/platform"
+	"github.com/weaveworks/flux/platform/rpc"
+)
+
+const (
+	timeout      = 5 * time.Second
+	presenceTick = 50 * time.Millisecond
+	encoderType  = nats.JSON_ENCODER
+
+	methodPing         = ".Platform.Ping"
+	methodAllServices  = ".Platform.AllServices"
+	methodSomeServices = ".Platform.SomeServices"
+	methodRegrade      = ".Platform.Regrade"
+)
+
+type NATS struct {
+	url string
+	// It's convenient to send (or request) on an encoding connection,
+	// since that'll do encoding work for us. When receiving though,
+	// we want to decode based on the method as given in the subject,
+	// so we use a regular connection and do the decoding ourselves.
+	snd *nats.EncodedConn
+	rcv *nats.Conn
+}
+
+var _ platform.MessageBus = &NATS{}
+
+func NewMessageBus(url string) (*NATS, error) {
+	conn, err := nats.Connect(url)
+	if err != nil {
+		return nil, err
+	}
+	encConn, err := nats.NewEncodedConn(conn, encoderType)
+	if err != nil {
+		return nil, err
+	}
+	return &NATS{
+		url: url,
+		rcv: conn,
+		snd: encConn,
+	}, nil
+}
+
+// Wait up to `timeout` for a particular instance to connect. Mostly
+// useful for synchronising during testing.
+func (n *NATS) AwaitPresence(instID flux.InstanceID, timeout time.Duration) error {
+	timer := time.After(timeout)
+	attempts := time.NewTicker(presenceTick)
+	defer attempts.Stop()
+
+	for {
+		select {
+		case <-attempts.C:
+			if err := n.Ping(instID); err == nil {
+				return nil
+			}
+		case <-timer:
+			return errors.New("presence timeout")
+		}
+	}
+}
+
+func (n *NATS) Ping(instID flux.InstanceID) error {
+	var response PingResponse
+	err := n.snd.Request(string(instID)+methodPing, ping{}, &response, timeout)
+	if err == nil {
+		err = maybeError(response.Error)
+	}
+	return err
+}
+
+type AllServicesResponse struct {
+	Services []platform.Service
+	Error    string
+}
+
+type SomeServicesResponse struct {
+	Services []platform.Service
+	Error    string
+}
+
+type RegradeResponse struct {
+	Result rpc.RegradeResult
+	Error  string
+}
+
+type ping struct{}
```

Comment:
Finding `struct{}{}` in a method call conveys an obvious meaning to my weary Go eyes: that this parameter is almost certainly an empty interface, and that it's being decoded-into or encoded-from, and that in this case it doesn't need to be anything in particular, because `struct{}` has no methods by definition. Finding `ping{}`, on the other hand, is an unknown: it opens the possibility that the parameter should satisfy some interface, which I'm now obliged to track down.

Do as you like.